### PR TITLE
Error on rest parameter with trailing comma

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes https://github.com/babel/babel/issues/7460#issuecomment-369452435

This will likely be annoying for many users, so we should make sure this definitely isn't going to be allowed by any future ES spec.

(MDN ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas#Trailing_commas_in_functions)

### CODE REVIEW

**Comment by ajafff** [comment]
TSLint's `trailing-comma` rule enforces trailing commas even after rest parameters. This used to work since TypeScript allowed it and removed it during transpiling (to a target other than ESNext).

There's an option to forbid trailing commas in that position, but it's not enabled by default. That means this change will break a lot of projects.

**Comment by RyanCavanaugh** [comment]
Not to put too fine a point on it, but we're not responsible for projects enforcing that they have a syntax error, even if we mistakenly didn't flag that error

**Comment by loganfsmyth** [comment]
It looks like the same issue is present for array rest and object rest

```
var [arg, ...other,] = [1, 2, 3, 4];

var { arg2, ...rest, } = { arg2: 1};
```
also pass without error.

**Comment by ajafff** [comment]
And not only on binding patterns, but also in destructuring assignments:

```ts
({...foo,} = {});
[...bar,] = [];
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -26093,6 +26093,9 @@ namespace ts {
                     if (i !== (parameterCount - 1)) {
                         return grammarErrorOnNode(parameter.dotDotDotToken, Diagnostics.A_rest_parameter_must_be_last_in_a_parameter_list);
                     }
+                    if (parameters.hasTrailingComma) {
+                        return grammarErrorOnNode(parameter.dotDotDotToken, Diagnostics.A_rest_parameter_may_not_have_a_trailing_comma);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I would say that this is the wrong span to error on. Either error on the commma, or the parameter itself.

**Comment by andy-ms** [comment]
Noticed that tslint has a `esSpecCompliant` setting (palantir/tslint#3176). @adidahiya @ajafff I think this should be made the default and published so that people can fix their code before this PR breaks them.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -26067,11 +26069,9 @@ namespace ts {
             return grammarErrorOnNode(asyncModifier, Diagnostics._0_modifier_cannot_be_used_here, "async");
         }
 
-        function checkGrammarForDisallowedTrailingComma(list: NodeArray<Node>): boolean {
+        function checkGrammarForDisallowedTrailingComma(list: NodeArray<Node>, diag = Diagnostics.Trailing_comma_not_allowed): boolean {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Where is this optional parameter used?

Comment (andy-ms) [reply src/compiler/checker.ts]: See the four new calls added in this PR.

**Comment by mhegazy** [comment]
@andy-ms  please add a blurb about this change in https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes

**Comment by andy-ms** [comment]
@mhegazy Good to merge now that 2.8 is released?

**Comment by mhegazy** [comment]
👍

**Comment by zixia** [comment]
@ajafff I run into this issue with `TSLint`'s `trailing-comma` rule. 

> There's an option to forbid trailing commas in that position, but it's not enabled by default.

After some search, I did not find the option you mentioned, would be appreciate if you could post it here, thanks!

**Pull request merged**